windows/src/main/csharp/ch/cyberduck/ui/core/preferences/PackagedDataMigrator.cs

```diff
@@ -0,0 +1,102 @@
+﻿using Ch.Cyberduck.Core.Preferences;
+using org.apache.logging.log4j;
+using System;
+using System.IO;
+using Windows.Win32.UI.Shell;
+using static Windows.Win32.CorePInvoke;
+using static Windows.Win32.PInvoke;
+
+namespace Ch.Cyberduck.Core.Ui.Preferences;
+
+internal static class PackagedDataMigrator
+{
+    public static void Migrate()
+    {
+        // Logger is used for short time.
+        // Don't need to keep an instance alive longer than necessary
+        var logger = LogManager.getLogger(typeof(PackagedDataMigrator).FullName);
+
+        string path;
+        try
+        {
+            const KNOWN_FOLDER_FLAG FLAGS = KNOWN_FOLDER_FLAG.KF_FLAG_RETURN_FILTER_REDIRECTION_TARGET | KNOWN_FOLDER_FLAG.KF_FLAG_DONT_VERIFY;
+            path = SHGetKnownFolderPath(FOLDERID_RoamingAppData, FLAGS, default);
+        }
+        catch (Exception e)
+        {
+            logger.warn("Failure getting RoamingAppData packaged folder path", e);
+            return;
+        }
+
+        path = Path.Combine(path, Runtime.DataFolderName);
+        if (!Directory.Exists(path))
+        {
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("Skipping migration for non-existent data folder");
+            }
+
+            return;
+        }
+
+        DirectoryInfo migratedLocation = new(Path.Combine(path, "Migrated"));
+        migratedLocation.Create();
+        var targetPath = RoamingSupportDirectoryFinder.Local.getAbsolute();
+        foreach (var item in new[] { "Bookmarks", "History", "Profiles", "Sessions", "Transfers" })
+        {
+            DirectoryInfo directory = new(Path.Combine(path, item));
+            if (!directory.Exists)
+            {
+                continue;
+            }
+
+            DirectoryInfo migratedDirectory = new(Path.Combine(migratedLocation.FullName, item));
+            migratedDirectory.Create();
+            var copyTarget = Path.Combine(targetPath, item);
+            foreach (var file in directory.EnumerateFiles())
+            {
+                bool failedMigrate = false;
+                FileInfo targetFile = new(Path.Combine(copyTarget, file.Name));
+                try
+                {
+                    if (targetFile.Exists && targetFile.LastWriteTime >= file.LastWriteTime)
+                    {
+                        // already migrated, or newer on shared location
+                        // move file to migrated-location
+
+                        if (logger.isInfoEnabled())
+                        {
+                            logger.info($"Skipping {item}/{file.Name} - Exists and target is newer");
+                        }
+
+                        continue;
```

Comment:
Finally block is executed regardless.

Reply:
Right, because following cases are covered:
- Copy from Packaged to Shared Location, but file exists in Shared Location already
  - Don't try to copy that file _every single time_ this migration is performed
- Copy from Packaged to Shared Location, but copy process failed
  - Don't move the migration-file, and try again next time
- Copy from Packaged to Shared Location, but file exists in Shared Location, and is older
  - Only move file away in case of successfully copying to shared location

Reply:
Exists in Shared, and timestamp equal => move to Migrated location
Exists in Shared, packaged newer => move to Migrated location if successful
Doesn't exist in Shared => move to Migrated location if successful

If copying to Shared location fails, don't move file to Migrated location, thus retrying the next time Cyberduck is started.